addons/product/models/product_product.py

```diff
@@ -184,27 +184,53 @@ def init(self):
         self.env.cr.execute("CREATE UNIQUE INDEX IF NOT EXISTS product_product_combination_unique ON %s (product_tmpl_id, combination_indices) WHERE active is true"
             % self._table)
 
-    @api.constrains('barcode')
-    def _check_barcode_uniqueness(self):
-        """ With GS1 nomenclature, products and packagings use the same pattern. Therefore, we need
-        to ensure the uniqueness between products' barcodes and packagings' ones"""
-        all_barcode = [b for b in self.mapped('barcode') if b]
-        domain = [('barcode', 'in', all_barcode)]
+    def _get_barcodes_by_company(self):
+        barcodes_by_company = defaultdict(list)
+
+        for product in self:
+            if product.barcode:
+                barcodes_by_company[product.company_id.id].append(product.barcode)
+
+        return barcodes_by_company
+
+    def _get_barcode_search_domain(self, barcodes_within_company, company_id):
+        domain = [('barcode', 'in', barcodes_within_company)]
+        if company_id:
+            domain.append(('company_id', 'in', (False, company_id)))
+        return domain
+
+    def _check_duplicated_product_barcodes(self, barcodes_within_company, company_id):
+        domain = self._get_barcode_search_domain(barcodes_within_company, company_id)
         matched_products = self.sudo().search(domain, order='id')
-        if len(matched_products) > len(all_barcode):  # It means that you find more than `self` -> there are duplicates
-            products_by_barcode = defaultdict(list)
-            for product in matched_products:
-                products_by_barcode[product.barcode].append(product)
-
-            duplicates_as_str = "\n".join(
-                _("- Barcode \"%s\" already assigned to product(s): %s", barcode, ", ".join(p.display_name for p in products))
-                for barcode, products in products_by_barcode.items() if len(products) > 1
+
```

Comment:
optimize the only line that didn't change
matched_products = self.sudo()._search(domain)

Reply:
Here I used `read_group` combined with `search`. I understand that the purpose was to improve performance, and I think it works in this regards, as we don't perform the search unless we have a list with more than one item. I tried to use `browse` instead of `search`, but there were some problems (full error log is a bit too long to post here).

Please let me know if this solution is good enough or if you want to discuss it further @amoyaux .

Reply:
@amoyaux  I guess we can keep the current solution that uses `search` but in a much more limited way (only searching if there are duplicated barcodes)?